tests/test_dataset.py

```diff
@@ -717,7 +726,10 @@ def test_tiled_tiff_read_and_write():
         pattern="{xxx}_{yyy}_{zzz}.tif",
     )
 
-    mag = tiled_tiff_ds.add_layer("color", Layer.COLOR_TYPE).add_mag("1")
+    mag = cast(
+        TiledTiffMagDataset,
+        tiled_tiff_ds.add_layer("color", Layer.COLOR_TYPE).add_mag("1"),
```

Comment:
What would be the actual return type of `add_mag`? I'm just wondering why we need to cast here, but probably fine :+1:

Reply:
There are three different types of datasets (`WKDataset`, `TiffDataset`, `TiledTiffDataset`), each of these datasets has a corresponding "MagDataset" (`WKMagDataset`, `TiffMagDataset`, `TiledTiffMagDataset`) (with `MagDataset` being the base class). 
`TiledTiffMagDataset` has an extra method, to access the tiles by an index. This is not part of `MagDataset` and the only case where a subclass defines an additional function. 
The function `add_mag` of `Layer` returns the corresponding MagDataset, but the typing hint is only `MagDataset` (not e.g. `TiledTiffMagDataset`) . Therefore, the typechecker would complain about `assert mag.get_tile(1, 1, 6).shape == (1, 32, 64, 1)` if the type is not casted.

Reply:
Ok, this is fine for now :+1: If we need specialized corresponding classes more often, we can use generics for this. E.g. defining
```
MagT = TypeVar('MagT')
class Layer(Generic[MagT]):
    add_layer(…) -> MagT:
        …

class TiledTiffLayer(Layer[TiledTiffMagDataset]):
    …
```